src/FindIntrinsics.cpp

```diff
@@ -292,6 +392,49 @@ class FindIntrinsics : public IRMutator {
             return mutate(result);
         }
 
+        if (op->type.is_int_or_uint() && op->type.bits() > 8) {
+            // Look for widen_right_mul intrinsics.
+            // Yes we do an duplicate code, but we want to check the op->type.code() first,
+            // and the opposite as well.
+            for (halide_type_code_t code : {op->type.code(), halide_type_uint, halide_type_int}) {
+                Type narrow = op->type.narrow().with_code(code);
+                Expr narrow_a = lossless_cast(narrow, a);
+                Expr narrow_b = lossless_cast(narrow, b);
+
+                // This case should have been handled by the above check for widening_mul.
+                internal_assert(!(narrow_a.defined() && narrow_b.defined()))
+                    << "find_intrinsics failed to find a widening_mul: " << a << " + " << b << "\n";
+
+                if (narrow_a.defined()) {
+                    Expr result;
+                    if (b.type().code() != narrow_a.type().code()) {
+                        // Need to do a safe reinterpret.
+                        Type t = b.type().with_code(code);
+                        result = widen_right_mul(reinterpret(t, b), narrow_a);
```

Comment:
(This is a follow-up to another comment from Steven regarding failures in C++ backend).

I am a bit confused by reinterpret here, do I understand it correctly that it might turn wild_i32x * u32(wild_u16x) into widen_right_mul(wild_u32x, wild_u16x)? 

The problem we are seeing is in the Xtensa backend (not exactly a C++ backend, but it's derived from it; also, lives in a separate branch, so it's a bit harder to keep track of), we currently don't handle this intrinsic, so at the code generation stage we will call lower_intrinsic once we encounter widen_right_mul. As a result, the following will happen:

1) we start with the expession like wild_i32x * widen(wild_u16x)
2) it gets transformed into widen_right_mul(wild_u32x, wild_u16x)
3) it gets lowered back to wild_u32x * widen(wild_u32x) [notice that left operand became unsingned]

If that's correct then the input of the 1) is not equiualent to the output of the 3), which seems a bit problematic? Is this transformation correct from numerical point of view (I guess depends on the actual implementation)? The specific problem we see in the Xtensa backend, is that Xtensa doesn't seem to have an intrinsic for multiplication of two wild_u32x vectors, but does have intrinsics for wild_i32x * wild_i32x and wild_i32x * wild_u32x (I know it's a bit weird and I can try to find out more details about it, but it may take some time).

Reply:
Numerically, the expressions are still equivalent, integer multiplication is the same for signed or unsigned arguments. 

I see your issue though, and I see two possible solutions:

1) in lowering intrinsics, specifically look for the pattern reinterpret(widen_right_op(reinterpret(a), b)), and lower it without the reinterprets (this seems messy)

2) Xtensa should specifically pattern match widen_right_mul(reinterpret(i32), u16) and use the wild_i32 * wild_u32 op that you mentioned

I believe that the letter is better, what do you think? Feasibly both could be implemented.

Reply:
I guess we could also change the design of the intrinsics to be
a op cast(a.type(), b) 
I think there was some reason that we chose not to do that initially. @abadams do you remember why?

Reply:
yes, 1) would be better in my opinion, but seems to be difficult to implement due to the outer reinterpret(), so probably not worth the effort or complexity.

I certainly can do 2), this should be pretty straigtforward. I was concerned that expressions are not equiualent after find_intrinsics -> lower_intrinsic, but it sounds it should be good numeric-wise.

I think, if this is the only issue we see in Google testing then it should be fine to merge in and I can address the  issue before updating Halide in Google.

Reply:
I'd prefer we address the Xtensa issue first, so that we can complete a test of this change in Google before landing. (Currently there are a lot of false-positive failures in the test due to this.)

Reply:
That's cool with me, I'll look into it.

Reply:
@vksnk I realize now that I should have questioned deeper on the Xtensa multiplication intrinsics - because multiplication is not parameterized by sign, shouldn’t the intrinsic that you mentioned being used for i32 x i32 multiplication be used for any 32-bit integer multiplication? 

That being said, I’m a little unsure what the i32 x u32 multiplication is used for. Is that a widening multiply by chance?

Reply:
And to address your concern about find_intrinsics -> lower_intrinsics not matching perfectly - unfortunately, that is already the case for most (possibly all?) of the intrinsics, though I believe these are the only intrinsics that will add reinterprets